# Cleaning notebook 

In [1]:
import numpy as np
import pandas as pd 

In [2]:
pd.set_option('display.max_columns', 50)
pd.set_option('display.max_rows', 80)
pd.set_option('display.width', 1000)

The purpose of this notebook is to clean the datasets I use during this project.

It is divided by section, one section for every dataset used. 

## Contributions :

### Reading data : 

In [3]:
data_ecolo = pd.read_csv('Donnees/data_contributions/LA_TRANSITION_ECOLOGIQUE.csv',error_bad_lines=False)
data_serv_pub = pd.read_csv('Donnees/data_contributions/ORGANISATION_DE_LETAT_ET_DES_SERVICES_PUBLICS.csv',error_bad_lines=False)
data_fisc = pd.read_csv('Donnees/data_contributions/LA_FISCALITE_ET_LES_DEPENSES_PUBLIQUES.csv',error_bad_lines=False)
data_dem = pd.read_csv('Donnees/data_contributions/DEMOCRATIE_ET_CITOYENNETE.csv',error_bad_lines=False)

C:\Users\adrie\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
b'Skipping line 6140: expected 44 fields, saw 62\n'
C:\Users\adrie\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (7,10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
C:\Users\adrie\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


###  Cleaning it :

On détecte qu'un des codes postaux est en fait une réponse à une question dans la base de données sur les services publiques : 

(Et que tout est mal indexé)

In [4]:
data_serv_pub[pd.to_numeric(data_serv_pub.authorZipCode, errors='coerce').isnull()]

id reference                                    title            createdAt          publishedAt updatedAt  trashed trashedStatus                                           authorId                                         authorType                                      authorZipCode QUXVlc3Rpb246MTY5 - Que pensez-vous de l'organisation de l'Etat et des administrations en France ? De quelle manière cette organisation devrait-elle évoluer ? QUXVlc3Rpb246MTcw - Selon vous, l'Etat doit-il aujourd'hui transférer de nouvelles missions aux collectivités territoriales ? QUXVlc3Rpb246MTcx - Si oui, lesquelles ? QUXVlc3Rpb246MjA0 - Estimez-vous avoir accès aux services publics dont vous avez besoin ? QUXVlc3Rpb246MTcy - Si non, quels types de services publics vous manquent dans votre territoire et qu'il est nécessaire de renforcer ?  \
6096  UHJvcG9zYWw6ODBlMjFiOTMtMWYwZC0xMWU5LTk0ZDItZm...    4-6706  Ne pas privatiser les barrages français  2019-01-23 13:50:46  2019-01-23 13:50:46       NaN    False           NaN  VXNlcjoxZT, ils ne sont pas responsables jurid...  Déjà en responsabilisant un peu les fonctionna...  Moi non, mais mon frère et plusieurs amis, oui...                                      Non concernée                                                                                                              J'ai été inscrite à Pôle Emploi et franchement...                                                                                                       Non concernée  Pendant des années, j'ai eu des soucis récurre...                                         Pas de soucis particulier vis-à-vis de la MDPH...                                                                                       

     QUXVlc3Rpb246MTc0 - Quels nouveaux services ou quelles démarches souhaitez-vous voir développées sur Internet en priorité ? QUXVlc3Rpb246MTc1 - Avez-vous déjà utilisé certaines de ces nouvelles formes de services publics ? QUXVlc3Rpb246MTc2 - Si oui, en avez-vous été satisfait ? QUXVlc3Rpb246MTc3 - Quelles améliorations préconiseriez-vous ? QUXVlc3Rpb246MTc4 - Quand vous pensez à l'évolution des services publics au cours des dernières années, quels sont ceux qui ont évolué de manière positive ? QUXVlc3Rpb246MTc5 - Quels sont les services publics qui doivent le plus évoluer selon vous ? QUXVlc3Rpb246MTgw - Connaissez-vous le "droit à l'erreur", c'est-à-dire le droit d'affirmer votre bonne foi lorsque vous faites un erreur dans vos déclarations ? QUXVlc3Rpb246MTgx - Si oui, avez-vous déjà utilisé ce droit à l'erreur ? QUXVlc3Rpb246MTgy - Si oui, à quelle occasion en avez-vous fait usage ?  \
6096                                                NaN                                                                                                                         NaN                                                                                                NaN                                                      NaN                                                            NaN                                                                                                                                                          NaN                                                                                          NaN                                                                                                                                                               NaN                                                                      NaN                        

     QUXVlc3Rpb246MTgz - Pouvez-vous identifier des règles que l'administration vous a déjà demandé d'appliquer et que vous avez jugées inutiles ou trop complexes ? QUXVlc3Rpb246MTg0 - Faut-il donner plus d'autonomie aux fonctionnaires de terrain ? QUXVlc3Rpb246MTg1 - Si oui, comment ? QUXVlc3Rpb246MTg2 - Faut-il revoir le fonctionnement et la formation de l'administration ? QUXVlc3Rpb246MTg3 - Si oui, comment ? QUXVlc3Rpb246MTg4 - Comment l'Etat et les

**On supprime cette ligne dans le ficher des contributions "Organisation de l'état et des service publics"**

In [5]:
data_serv_pub.drop(6096,inplace=True)

In [8]:
data_serv_pub=data_serv_pub.drop(['id'],axis=1)
data_dem=data_dem.drop(['id'],axis=1)
data_fisc=data_fisc.drop(['id'],axis=1)

### Saving it 

In [9]:
data_ecolo.to_csv('Donnees_clean/contributions/data_ecolo.csv',index=False)
data_serv_pub.to_csv('Donnees_clean/contributions/data_serv_pub.csv', index=False)
data_fisc.to_csv('Donnees_clean/contributions/data_fisc.csv',index=False)
data_dem.to_csv('Donnees_clean/contributions/data_dem.csv',index=False)

## Zip code : (OK)

### Reading data : 

In [32]:
Data_Zip_Code = pd.read_csv('Donnees/data_code_postal/laposte_hexasmal.csv',';')
Data_Zip_Code.head()

Code_commune_INSEE               Nom_commune  Code_postal      Libelle_acheminement Ligne_5              coordonnees_gps
0              90093                SERMAMAGNY        90300                SERMAMAGNY     NaN    47.687801557,6.8309146345
1              91093         BOULLAY LES TROUX        91470         BOULLAY LES TROUX     NaN  48.6753515056,2.04828313772
2              91100                  BOUVILLE        91880                  BOUVILLE     NaN   48.4326483441,2.2783856422
3              91129                     CERNY        91590                     CERNY     NaN  48.4859798517,2.31068283872
4              91184  COURDIMANCHE SUR ESSONNE        91720  COURDIMANCHE SUR ESSONNE     NaN   48.418031424,2.36344067611

### Cleaning data

In [33]:
Data_Zip_Code.isna().sum()

Code_commune_INSEE          0
Nom_commune                 0
Code_postal                 0
Libelle_acheminement        0
Ligne_5                 35944
coordonnees_gps           269
dtype: int64

Il y a des nans dans cette base de données à la place des coordonnées gps. 

On ne prend en compte que les codes postaux situés en France metropolitaine : 

In [34]:
#On ne selectionne que les codes postaux situés entre 1000 et 96000 :
Data_Zip_Code=Data_Zip_Code.loc[Data_Zip_Code['Code_postal']>=1000]
Data_Zip_Code=Data_Zip_Code.loc[Data_Zip_Code['Code_postal']<96000]

In [35]:
Data_Zip_Code.isna().sum()

Code_commune_INSEE          0
Nom_commune                 0
Code_postal                 0
Libelle_acheminement        0
Ligne_5                 35755
coordonnees_gps             2
dtype: int64

On se rend compte qu'il y a beacoup moins de nans si on ne prend que les codes postaux de France metropolitaine. 

In [36]:
#On Organise les données en les triant selon les valeurs du code postal : 
Data_Zip_Code.sort_values(by=['Code_postal'],inplace=True)
#On reset l'index pour avoir toujours 1,2,...n-1 : 
Data_Zip_Code.reset_index(drop=True,inplace=True)

In [38]:
Data_Zip_Code.head()

Code_commune_INSEE         Nom_commune  Code_postal Libelle_acheminement Ligne_5              coordonnees_gps
0              01344  ST DENIS LES BOURG         1000   ST DENIS LES BOURG     NaN  46.2111998453,5.18462806771
1              01053     BOURG EN BRESSE         1000      BOURG EN BRESSE     NaN  46.2051520382,5.24602125501
2              01053     BOURG EN BRESSE         1000      BOURG EN BRESSE    BROU  46.2051520382,5.24602125501
3              01165        FRANCHELEINS         1090         FRANCHELEINS     NaN  46.0735946256,4.81293575559
4              01225               LURCY         1090                LURCY     NaN   46.061730034,4.78075207026

In [39]:
#On remplace les coordonnées GPS par les coordonnées GPS précédentes dans la colonne : 
Data_Zip_Code['coordonnees_gps'].fillna(method='pad', inplace=True)

In [40]:
Data_Zip_Code.isna().sum()

Code_commune_INSEE          0
Nom_commune                 0
Code_postal                 0
Libelle_acheminement        0
Ligne_5                 35755
coordonnees_gps             0
dtype: int64

On se fiche des ligne_5 qui continent des nans et qui correspond en fait à une sorte de lieu dit.

On supprime cette colonne.

In [41]:
Data_Zip_Code.drop(columns='Ligne_5',inplace=True)

In [42]:
Data_Zip_Code.head()

Code_commune_INSEE         Nom_commune  Code_postal Libelle_acheminement              coordonnees_gps
0              01344  ST DENIS LES BOURG         1000   ST DENIS LES BOURG  46.2111998453,5.18462806771
1              01053     BOURG EN BRESSE         1000      BOURG EN BRESSE  46.2051520382,5.24602125501
2              01053     BOURG EN BRESSE         1000      BOURG EN BRESSE  46.2051520382,5.24602125501
3              01165        FRANCHELEINS         1090         FRANCHELEINS  46.0735946256,4.81293575559
4              01225               LURCY         1090                LURCY   46.061730034,4.78075207026

In [43]:
Data_Zip_Code.isna().sum()

Code_commune_INSEE      0
Nom_commune             0
Code_postal             0
Libelle_acheminement    0
coordonnees_gps         0
dtype: int64

### Saving it 

In [46]:
Data_Zip_Code.to_csv('Donnees_clean/data_code_postal/Data_Zip_Code.csv')

## Elections :

### Reading data : 

In [47]:
elections_pres=pd.read_csv('Donnees/presidentielles/Presidentielle_2017_Resultats_Communes_T1_clean_def.csv',error_bad_lines=False)

C:\Users\adrie\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (0,1) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [49]:
elections_pres.head()

CodeInsee CodeDepartement Département                  Commune  Inscrits  Abstentions  Abstentions_ins  Votants  Votants_ins  Blancs  Blancs_ins  Blancs_vot  Nuls  Nuls_ins  Nuls_vot  Exprimés  Exprimés_ins  Exprimés_vot  LE PEN  MÉLENCHON  MACRON  FILLON  DUPONT-AIGNAN  LASSALLE  HAMON  ...  POUTOU  ARTHAUD  CHEMINADE  LE PEN.ins  MÉLENCHON.ins  MACRON.ins  FILLON.ins  DUPONT-AIGNAN.ins  LASSALLE.ins  HAMON.ins  ASSELINEAU.ins  POUTOU.ins  ARTHAUD.ins  CHEMINADE.ins  LE PEN.exp  MÉLENCHON.exp  MACRON.exp  FILLON.exp  DUPONT-AIGNAN.exp  LASSALLE.exp  HAMON.exp  ASSELINEAU.exp  POUTOU.exp  ARTHAUD.exp  CHEMINADE.exp
0     01001              01         Ain  L'Abergement-Clémenciat       598           92        15.384615      506    84.615385       2    0.334448    0.395257     9  1.505017  1.778656       495     82.775920     97.826087     126         59     119     110             34         2     29  ...       4        4          2   21.070234       9.866221   19.899666   18.394649           5.685619      0.334448   4.849498        1.003344    0.668896     0.668896       0.334448   25.454545      11.919192   24.040404   22.222222           6.868687      0.404040   5.858586        1.212121    0.808081     0.808081       0.404040
1     01002              01         Ain    L'Abergement-de-Varey       209           25        11.961722      184    88.038278       6    2.870813    3.260870     2  0.956938  1.086957       176     84.210526     95.652174      48         33      37      34              6         0     13  ...       2        2          0   22.966507      15.789474   17.703349   16.267943           2.870813      0.000000   6.220096        0.478469    0.956938     0.956938       0.000000   27.272727      18.750000   21.022727   19.318182           3.409091      0.000000   7.386364        0.568182    1.136364     1.136364       0.000000
2     01004              01         Ain        Ambérieu-en-Bugey      8586         1962        22.851153     6624    77.148847     114    1.327743    1.721014    58  0.675518  0.875604      6452     75.145586     97.403382    1667       1412    1332    1084            346        60    344  ...      91       40          5   19.415327      16.445376   15.513627   12.625204           4.029816      0.698812   4.006522        0.826928    1.059865     0.465875       0.058234   25.836950      21.884687   20.644761   16.800992           5.362678      0.929944   5.331680        1.100434    1.410415     0.619963       0.077495
3     01005              01         Ain      Ambérieux-en-Dombes      1172          215        18.344710      957    81.655290      21    1.791809    2.194357     3  0.255973  0.313480       933     79.607509     97.492163     306        126     191     197             45         6     37  ...      10        5          0   26.109215      10.750853   16.296928   16.808874           3.839590      0.511945   3.156997        0.853242    0.853242     0.426621       0.000000   32.797428      13.504823   20.471597   21.114684           4.823151      0.643087   3.965702        1.071811    1.071811     0.535906       0.000000
4     01006              01         Ain                  Ambléon        99           20        20.202020       79    79.797980       2    2.020202    2.531646     0  0.000000  0.000000        77     77.777778     97.468354      18         19      15      14              4         1      3  ...       2        1          0   18.181818      19.191919   15.151515   14.141414           4.040404      1.010101   3.030303        0.000000    2.020202     1.010101       0.000000   23.376623      24.675325   19.480519   18.181818           5.194805      1.298701   3.896104        0.000000    2.597403     1.298701       0.000000

[5 rows x 51 columns]

In [54]:
elections_pres.isna().sum()

CodeInsee            0
CodeDepartement      0
Département          0
Commune              0
Inscrits             0
Abstentions          0
Abstentions_ins      1
Votants              0
Votants_ins          1
Blancs               0
Blancs_ins           1
Blancs_vot           2
Nuls                 0
Nuls_ins             1
Nuls_vot             2
Exprimés             0
Exprimés_ins         1
Exprimés_vot         2
LE PEN               0
MÉLENCHON            0
MACRON               0
FILLON               0
DUPONT-AIGNAN        0
LASSALLE             0
HAMON                0
ASSELINEAU           0
POUTOU               0
ARTHAUD              0
CHEMINADE            0
LE PEN.ins           1
MÉLENCHON.ins        1
MACRON.ins           1
FILLON.ins           1
DUPONT-AIGNAN.ins    1
LASSALLE.ins         1
HAMON.ins            1
ASSELINEAU.ins       1
POUTOU.ins           1
ARTHAUD.ins          1
CHEMINADE.ins        1
LE PEN.exp           8
MÉLENCHON.exp        8
MACRON.exp           8
FILLON.exp 

Il y a peu de nans.

###  Cleaning it :

### Saving it 

## twitter and #Elysée2017fr dataset :

### Reading data : 

Les fichiers sont en .tab donc il faut ajuster les paramètres de `pd.read_csv` :

In [2]:
dfprofiles=pd.read_csv('Donnees/data_elysee2017_tweets/profiles_annotations.tab', sep='\t')

In [3]:
dfprofiles.head()

FROM_USER_ID PROFILE_NATURE PARTY  MEDIA_PROFESSIONAL SEX
0    1264925822     individual    fi                 NaN   f
1    1265204928     individual    fi                 NaN   f
2    1265310602     individual    fi                 NaN   m
3    1265661672     individual    fi                 NaN   m
4    1267543572     individual    fi                 NaN   m

###  Cleaning it :

#### Données sur le parti politique : 

In [4]:
dfprofiles['PARTY'].isna().sum()

3401

Il n'y a que 3401 nans pour le parti politique sur les 22853 profils disponibles. 

Etant donné qu'il serait très compliqué de les estimer (sauf à la main) donc on peut les supprimer. 

In [5]:
dfprofiles.dropna(0,subset=['PARTY'],inplace=True)

#### Autres données : 

In [6]:
dfprofiles.MEDIA_PROFESSIONAL.unique()

array([nan,  1.])

On remplace les nan dans `MEDIA_PROFESSIONAL` par 0 
- S'il s'agit d'un media professionel : 1
- S'il s'agit d'un individu : 0

In [7]:
dfprofiles['MEDIA_PROFESSIONAL'].fillna(0,inplace=True)

In [8]:
print('nombre de medias dans le base : ', dfprofiles[dfprofiles['MEDIA_PROFESSIONAL']==1].shape[0])
print("nombre d'individus"," dans le base : ", dfprofiles[dfprofiles['MEDIA_PROFESSIONAL']==0].shape[0])

nombre de medias dans le base :  46
nombre d'individus  dans le base :  19406


**Il y a beaucoup plus d'invidus que de media, ce qui est rassurant compte tenu de l'utilisation que l'on veut faire**

In [9]:
dfprofiles.isna().sum()

FROM_USER_ID             0
PROFILE_NATURE           0
PARTY                    0
MEDIA_PROFESSIONAL       0
SEX                   5338
dtype: int64

Il reste 5338 nans dans les données. 
**On ne s'en occupe pas pour l'instant**

#### Nombre d'utilisateurs uniques  : 

In [10]:
dfprofiles['FROM_USER_ID'].nunique()

19452

### Saving it 

In [13]:
dfprofiles.to_csv('Donnees_clean/twitter/utilisateurs/profiles_annotations.csv',index=False)